src/Stratis.Bitcoin.Features.Consensus/Rules/ProvenHeaderRules/ProvenHeaderCoinstakeRule.cs

```diff
@@ -355,9 +357,19 @@ private UnspentOutputs CheckIfCoinstakeIsSpentOnAnotherChain(ProvenBlockHeader h
             }
 
             RewindData rewindData = this.PosParent.UtxoSet.GetRewindData(rewindDataIndex.Value).GetAwaiter().GetResult();
-            UnspentOutputs matchingUnspentUtxo = rewindData.OutputsToRestore
-                .Where((unspent, i) => (unspent.TransactionId == input.PrevOut.Hash) && (i == input.PrevOut.N))
-                .FirstOrDefault();
+
```

Comment:
I wonder if here we should also add one more check 
```
            UnspentOutputs matchingUnspentUtxo = null;
            foreach (UnspentOutputs unspent in rewindData.OutputsToRestore)
            {
                if (unspent.TransactionId == input.PrevOut.Hash)
                {
                    if (input.PrevOut.N < unspent.Outputs.Length)
                    {
                        if (unspent.Outputs[input.PrevOut.N] != null)
                        {
                            matchingUnspentUtxo = unspent;
                            break;
                        }
                    }
                }
            }
```

Reply:
I am not sure because later we will check the UTXO is not null when we validate the signature, this is a permanent problem so if we do this check here we will just continue to try this header because later we mark this as `InsufficientHeaderInformation` so I suggest either to throw another error code or to not add this check at all.

Reply:
if we have a rewindIndex for that output shouldn't we have a rewindData too?
if yes, than yes we need to check nullity and throw because it's spent

if we don't throw, then matchingUnspentUtxo could mean we have insufficient informations